File: experimental/infogain.py
import math

class Primal:
    
    primes = {}

    def __init__(self):
        pass
    
    """ Returns true iff n is prime."""
    @staticmethod
    def isprime(n):
        if n < 2:
            return False
        if n == 2:
            return True
        if n % 2 == 0:
            return False
        for i in range(3,int(math.sqrt(n) + 1), 2):
            if n % i == 0:
                return False
        return True

    """Return the nth prime number (0 = 2)."""
    @staticmethod
    def nthprime(n):
        # Use a memorized table.
        if n in Primal.primes:
            return Primal.primes[n]
        if n == 0 or n == 1:
            Primal.primes[0] = 2
            Primal.primes[1] = 3
            return Primal.primes[n]
        for i in range(len(Primal.primes), n+1):
            curr_prime = Primal.primes[i-1]
            # Find the next prime number and record it.
            summand = 2
            while not Primal.isprime(summand + curr_prime):
                summand += 2
            Primal.primes[i] = summand + curr_prime
        return Primal.primes[n]



class InfoGain:

    """Compute normalized entropy of a pmf (between 0 and 1)."""
    @staticmethod
    def normalized_entropy(counts):
        total = sum(counts)
        entropy = 0
        for i in range(0,len(counts)):
            freq = counts[i] / total
            if freq != 0:
                entropy -= freq * math.log2(freq)
        return entropy / math.log2(total)


    """Compute Shannon entropy of a pmf."""
    @staticmethod
    def entropy(counts):
        total = sum(counts)
        entropy = 0
        for i in range(0,len(counts)):
            freq = counts[i] / total
            if freq != 0:
                entropy += -freq * math.log2(freq)
        return entropy

    """Compute the salience (entropy-discounted "importance", from 0 to 1) of an index in a pmf."""
    @staticmethod
    def salience(counts, index):
        total = sum(counts)
        p_index = counts[index] / total
        salience = (1 - p_index) * (1-InfoGain.normalized_entropy(counts))
        return salience
    
class Stencil:

    # Pick one of these stencils, whose names should be self-explanatory.
    useNinePointStencil = True
    useEightPointStencil = False  # exclude the center
    useFivePointStencil = False
    useOnePointStencil = False

    reflectStencils = True  # False

    @staticmethod
    def apply_stencil(stencil, arr, i, j, base, operator):
        v = base
        for (x,y) in stencil:
            # Transform x and y here, since the first coordinate is
            # actually the row (y-coord) and the second is the column
            # (x-coord).
            v = operator(v, arr[i + y][j + x])
        return v # (v / len(stencil)) # FIXME?
        
    @staticmethod
    def stencil_computation(arr, operator, base):
        # Array is 2D
        nrows = len(arr)
        ncols = len(arr[0])
        # Make a new array of zeroes of the same size.
        new_arr = [[0 for c in range(0,ncols)] for r in range(0,nrows)]
        # Define stencils by their coordinates (x offset, y offset).

        assert Stencil.useNinePointStencil ^ \
        Stencil.useEightPointStencil ^ \
        Stencil.useFivePointStencil ^ \
        Stencil.useOnePointStencil, "Exactly one stencil choice should be used."
        
        if Stencil.useNinePointStencil:
            stencil = [(-1,-1), (-1,0), (-1,1), (0,-1), (0, 0), (0,1), (1,-1), (1,0), (1,1)]

        if Stencil.useEightPointStencil:
            stencil = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]
            
        if Stencil.useFivePointStencil:
            # cross or "plus" stencil (five-points)
            stencil = [(-1,0), (0,-1), (0, 0), (1,0), (0,1)]

        if Stencil.useOnePointStencil:
            # just the center itself.
            stencil = [(0,0)]
        
        # Define boundary condition stencils by clipping the stencil at
        # the boundaries (edges and then corners).
        # NOTE: we optionally REFLECT the stencil here so it is always the same size.

        stencil_right    = [(x,y) for (x,y) in stencil if x <= 0]
        stencil_left     = [(x,y) for (x,y) in stencil if x >= 0]
        stencil_top = [(x, y) for (x, y) in stencil if y >= 0]
        
        print("top was = " + str(stencil_top))

        stencil_bottom   = [(x,y) for (x,y) in stencil if y <= 0]
        stencil_topleft  = [(x,y) for (x,y) in stencil_top if x >= 0]
        stencil_topright = [(x,y) for (x,y) in stencil_top if x <= 0]
        stencil_bottomleft  = [(x,y) for (x,y) in stencil_bottom if x >= 0]
        stencil_bottomright = [(x,y) for (x,y) in stencil_bottom if x <= 0]

        if Stencil.reflectStencils:
            # "reflect" the stencil for edge cases.
            stencil_right    += [(-x,y) for (x,y) in stencil if x > 0]
            stencil_left     += [(-x,y) for (x,y) in stencil if x < 0]
            stencil_top      += [(x,-y) for (x,y) in stencil if y < 0]

            print("top = " + str(stencil_top))

            stencil_bottom   += [(x,-y) for (x,y) in stencil if y > 0]
            stencil_topleft += [(-x, y) for (x, y) in stencil_top if x < 0] + [(x, -y) for (x, y) in stencil_left if y < 0]
            print("length of topleft = " + str(len(stencil_topleft)))
            stencil_topright += [(-x,y) for (x,y) in stencil_top if x > 0] + [(x,-y) for (x,y) in stencil_right if y < 0]
            stencil_bottomleft  += [(-x,y) for (x,y) in stencil_bottom if x < 0] + [(x,-y) for (x,y) in stencil_left if y > 0]
            stencil_bottomright += [(-x,y) for (x,y) in stencil_bottom if x > 0] + [(x,-y) for (x,y) in stencil_right if y > 0]

        # Interior
        for i in range(1,ncols-1):
            for j in range(1,nrows-1):
                new_arr[i][j] = Stencil.apply_stencil(stencil, arr, i, j, base, operator)
        # Edges
        ## Top and bottom
        for j in range(1,ncols-1):
            new_arr[0][j]         = Stencil.apply_stencil(stencil_top, arr, 0, j, base, operator)
            new_arr[nrows-1][j]   = Stencil.apply_stencil(stencil_bottom, arr, nrows-1, j, base, operator)
        ## Left and right
        for i in range(1,nrows-1):
            new_arr[i][0]         = Stencil.apply_stencil(stencil_left, arr, i, 0, base, operator)
            new_arr[i][ncols-1]   = Stencil.apply_stencil(stencil_right, arr, i, ncols-1, base, operator)
        # Corners
        new_arr[0][0]             = Stencil.apply_stencil(stencil_topleft, arr, 0, 0, base, operator)
        new_arr[0][ncols-1]       = Stencil.apply_stencil(stencil_topright, arr, 0, ncols-1, base, operator)
        new_arr[nrows-1][0]       = Stencil.apply_stencil(stencil_bottomleft, arr, nrows-1, 0, base, operator)
        new_arr[nrows-1][ncols-1] = Stencil.apply_stencil(stencil_bottomright, arr, nrows-1, ncols-1, base, operator)
        return new_arr



class Encoder:

    """Substitute each MATRIX entry with a (fixed) prime number."""
    @staticmethod
    def primal_matrix(mat):
        prime_array = {}
        nth = 0
        ret_mat = []
        for arr in mat:
            ret_array = []
            for i in arr:
                if i not in prime_array:
                    prime = Primal.nthprime(nth)
                    prime_array[i] = prime
                    nth += 1
                ret_array += [prime_array[i]]
            ret_mat.append(ret_array)
        return ret_mat


    """Substitute each MATRIX entry with a (fixed) 'one-hot' number."""
    @staticmethod
    def onehot_matrix(mat):
        onehot_array = {}
        nth = 0
        ret_mat = []
        for arr in mat:
            ret_array = []
            for i in arr:
                if i not in onehot_array:
                    onehot = 1 << nth
                    onehot_array[i] = onehot
                    nth += 1
                ret_array += [onehot_array[i]]
            ret_mat.append(ret_array)
        return ret_mat

# print(primal_array([1,2,3,4,5,1,2,3,4,5,1]))

def plus(x,y):
    return x+y

def times(x,y):
    return x*y

def bitor(x,y):
    return x | y

from collections import Counter

# arr = [[1,1,1],[1,2,1],[1,1,1]]
# arr = [[1,1,2,1],[1,1,1,1],[1,1,1,1],[1,1,1,1]]

# Use one-hot encoding.

#print(normalized_entropy([1000,1]))
print(InfoGain.salience([1000,1], 1))
print(InfoGain.salience([1000,1], 0))
print(InfoGain.salience([1,1], 0))
print(InfoGain.salience([1,1,1], 0))
print("trying a range:")
s = 0
for x in range(0,4):
    v = InfoGain.salience([1,10,1,1], x)
    s += v
    print(v)
print("--- : sum = " + str(s))
print(InfoGain.salience([1,20000], 0))
print(InfoGain.salience([1,2], 0))
print(InfoGain.salience([1,2,2], 0))
print(InfoGain.salience([1,2,2,2,2], 0))
print(InfoGain.salience([1,2,2,2,2,2,2,2,2], 0))

arr = [[1 for i in range(0,10)] for i in range(0,10)]
arr[3][5] = 14
arr[7][8] = 99

#arr = [[1,1,1,1,1],[1,1,1,1,1],[1,1,1,1,1],[1,1,1,1,1],[1,1,1,1,1]]
#arr = [[1,1,1,1,1],[2,2,2,2,2],[3,3,3,3,3],[4,4,4,4,4],[5,5,5,5,5]]
# arr = [[1,4,4,1,1],[1,1,1,1,1],[1,1,1,1,1],[1,1,8,1,1],[1,1,1,1,1]]
#arr = [[1,1,1,1,1],[1,1,8,1,1],[1,1,1,1,1],[1,1,1,1,1],[1,1,1,1,1]]
#arr = [[1,2,3,4,1],[1,2,8,3,1],[4,5,3,2,4],[5,2,3,4,1],[1,3,4,2,1]]
#arr = [[1,2,2,2,2],[2,2,8,2,2],[2,2,2,2,4],[2,2,2,1,1],[2,2,2,2,1]]
#arr = [[1,4,4,2,3],[5,6,7,100,9],[10,11,12,13,14],[15,16,8,17,18],[19,20,21,22,23]]

usePrimeEncoding =  False # True # False
useOneHotEncoding = False
useNoEncoding = True

assert usePrimeEncoding ^ useOneHotEncoding ^ useNoEncoding, "Exactly one encoding should be used."

# Use n-th prime encoding.

if usePrimeEncoding:
    #arr = [[1,2,2,1,1],[1,1,1,1,1],[1,1,1,1,1],[1,1,3,1,1],[1,1,1,1,1]]
    prime_encoding = Encoder.primal_matrix(arr)
    arr = prime_encoding
    print("prime encoding = " + str(arr))
    arr = Stencil.stencil_computation(arr, times, 1)

if useOneHotEncoding:
    arr = Encoder.onehot_matrix(arr)
    print("onehot = " + str(arr))
    arr = Stencil.stencil_computation(arr, bitor, 0)

if useNoEncoding:
    arr = Stencil.stencil_computation(arr, times, 1)

    
# arr = Stencil.stencil_computation(arr, bitor, 0)
# arr = Stencil.stencil_computation(arr, plus, 0)
z=arr
print(z)

# Compute salience for the whole array, post stencil.
# Flatten array.
flat_z = [item for sublist in z for item in sublist]
# Get counts.
counts = Counter(flat_z)
# Convert to an array
keys = list(counts.keys())
arr = list(counts.values())
print("arr = " + str(arr))
salience_array = [(keys[i],InfoGain.salience(arr, i)) for i in range(0,len(arr))]
print("salience array = " + str(salience_array))
#print(counts)

for i in range(0, len(z)):
    for j in range(0, len(z[0])):
        pass
    

File: experimental/infogain.js
"use strict";
exports.__esModule = true;
var InfoGain = /** @class */ (function () {
    function InfoGain() {
    }
    InfoGain.normalized_entropy = function (counts) {
        var total = counts.reduce(function (a, b) { return a + b; }, 0);
        var entropy = 0;
        for (var i = 0; i < counts.length; i++) {
            var freq = counts[i] / total;
            if (freq !== 0) {
                entropy -= freq * Math.log2(freq);
            }
        }
        return entropy / Math.log2(total);
    };
    InfoGain.salience = function (counts, index) {
        var total = counts.reduce(function (a, b) { return a + b; }, 0);
        var p_index = counts[index] / total;
        var salience = (1 - p_index) * (1 - InfoGain.normalized_entropy(counts));
        return salience;
    };
    return InfoGain;
}());
exports.InfoGain = InfoGain;
var Stencil = /** @class */ (function () {
    function Stencil() {
    }
    Stencil.initialize = function () {
        if (!Stencil.initialized) {
            //  Define boundary condition stencils by clipping the stencil at
            // the boundaries(edges and then corners).
            // NOTE: we optionally REFLECT the stencil here so it is always the same size.
            Stencil.stencil_right = Stencil.stencil.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x <= 0);
            });
            Stencil.stencil_left = Stencil.stencil.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x >= 0);
            });
            Stencil.stencil_top = Stencil.stencil.filter(function (_a) {
                var _ = _a[0], y = _a[1];
                return (y >= 0);
            });
            Stencil.stencil_bottom = Stencil.stencil.filter(function (_a) {
                var _ = _a[0], y = _a[1];
                return (y <= 0);
            });
            Stencil.stencil_topleft = Stencil.stencil_top.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x >= 0);
            });
            Stencil.stencil_topright = Stencil.stencil_top.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x <= 0);
            });
            Stencil.stencil_bottomleft = Stencil.stencil_bottom.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x >= 0);
            });
            Stencil.stencil_bottomright = Stencil.stencil_bottom.filter(function (_a) {
                var x = _a[0], _ = _a[1];
                return (x <= 0);
            });
            if (Stencil.reflectStencils) {
                var reflected = void 0;
                // Right
                reflected = Stencil.stencil.filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x > 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                });
                Stencil.stencil_right = Stencil.stencil_right.concat(reflected);
                console.log('stencil right = ' + JSON.stringify(Stencil.stencil_right) + ', length = ' + Stencil.stencil_right.length);
                // Left
                reflected = Stencil.stencil.filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x < 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                });
                Stencil.stencil_left = Stencil.stencil_left.concat(reflected);
                console.log('stencil left = ' + JSON.stringify(Stencil.stencil_left) + ', length = ' + Stencil.stencil_left.length);
                // Top
                reflected = Stencil.stencil.filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y < 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                });
                Stencil.stencil_top = Stencil.stencil_top.concat(reflected);
                console.log('stencil top = ' + JSON.stringify(Stencil.stencil_top) + ', length = ' + Stencil.stencil_top.length);
                // Bottom
                reflected = Stencil.stencil.filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y > 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                });
                Stencil.stencil_bottom = Stencil.stencil_bottom.concat(reflected);
                console.log('stencil bottom = ' + JSON.stringify(Stencil.stencil_bottom) + ', length = ' + Stencil.stencil_bottom.length);
                // Top left
                // stencil_topleft += [(-x, y) for (x, y) in stencil_top if x < 0]+[(x, -y) for (x, y) in stencil_left if y < 0]
                reflected = Stencil.stencil_top.filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x < 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                });
                Stencil.stencil_topleft = Stencil.stencil_topleft.concat(reflected);
                reflected = Stencil.stencil_left.filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y < 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                });
                Stencil.stencil_topleft = Stencil.stencil_topleft.concat(reflected);
                // Remove [1,1]
                Stencil.stencil_topleft.splice(Stencil.stencil_topleft.findIndex(function (o) { return JSON.stringify(o) === JSON.stringify([1, 1]); }), 1);
                console.log('stencil top left = ' + JSON.stringify(Stencil.stencil_topleft) + ', length = ' + Stencil.stencil_topleft.length);
                // Top right
                // stencil_topright += [(-x, y) for (x, y) in stencil_top if x > 0]+[(x, -y) for (x, y) in stencil_right if y < 0]
                reflected = Stencil.stencil_top.filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x > 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                });
                Stencil.stencil_topright = Stencil.stencil_topright.concat(reflected);
                reflected = Stencil.stencil_right.filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y < 0);
                }).map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                });
                Stencil.stencil_topright = Stencil.stencil_topright.concat(reflected);
                // Remove [-1,1]
                Stencil.stencil_topright.splice(Stencil.stencil_topright.findIndex(function (o) { return JSON.stringify(o) === JSON.stringify([-1, 1]); }), 1);
                console.log('stencil top right = ' + JSON.stringify(Stencil.stencil_topright) + ', length = ' + Stencil.stencil_topright.length);
                // Bottom left
                // stencil_bottomleft += [(-x, y) for (x, y) in stencil_bottom if x < 0]+[(x, -y) for (x, y) in stencil_left if y > 0]
                reflected = Stencil.stencil_bottom.map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                }).filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x > 0);
                });
                Stencil.stencil_bottomleft = Stencil.stencil_bottomleft.concat(reflected);
                reflected = Stencil.stencil_left.map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                }).filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y < 0);
                });
                Stencil.stencil_bottomleft = Stencil.stencil_bottomleft.concat(reflected);
                // Remove [1,-1]
                Stencil.stencil_bottomleft.splice(Stencil.stencil_bottomleft.findIndex(function (o) { return JSON.stringify(o) === JSON.stringify([1, -1]); }), 1);
                console.log('stencil bottom left = ' + JSON.stringify(Stencil.stencil_bottomleft) + ', length = ' + Stencil.stencil_bottomleft.length);
                // Bottom right
                // stencil_bottomright += [(-x, y) for (x, y) in stencil_bottom if x > 0]+[(x, -y) for (x, y) in stencil_right if y > 0]
                reflected = Stencil.stencil_bottom.map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [-x, y];
                }).filter(function (_a) {
                    var x = _a[0], _ = _a[1];
                    return (x < 0);
                });
                Stencil.stencil_bottomright = Stencil.stencil_bottomright.concat(reflected);
                reflected = Stencil.stencil_right.map(function (_a) {
                    var x = _a[0], y = _a[1];
                    return [x, -y];
                }).filter(function (_a) {
                    var _ = _a[0], y = _a[1];
                    return (y < 0);
                });
                Stencil.stencil_bottomright = Stencil.stencil_bottomright.concat(reflected);
                // Remove [-1,-1]
                Stencil.stencil_bottomright.splice(Stencil.stencil_bottomright.findIndex(function (o) { return JSON.stringify(o) === JSON.stringify([-1, -1]); }), 1);
                console.log('stencil bottom right = ' + JSON.stringify(Stencil.stencil_bottomright) + ', length = ' + Stencil.stencil_bottomright.length);
            }
            Stencil.initialized = true;
        }
    };
    Stencil.apply_stencil = function (stencil, arr, i, j, base, operator) {
        console.assert(stencil.length === Stencil.stencil.length);
        var v = base;
        for (var ind = 0; ind < stencil.length; ind++) {
            var _a = stencil[ind], x = _a[0], y = _a[1];
            // console.log('[x,y] = [' + x + ',' + y + ']');
            // Transform x and y here, since the first coordinate is
            // actually the row(y - coord) and the second is the column
            // (x - coord).
            v = operator(v, arr[i + y][j + x]);
        }
        return v;
    };
    Stencil.stencil_computation = function (arr, operator, base) {
        Stencil.initialize();
        // Make a new matrix ("new_arr") of the same size as arr, initialized with zeros.
        var nrows = arr.length;
        var ncols = arr[0].length;
        var new_arr = Array(nrows).fill(0).map(function () { return Array(ncols).fill(0); });
        // Interior
        for (var i = 1; i < ncols - 1; i++) {
            for (var j = 1; j < nrows - 1; j++) {
                new_arr[i][j] = Stencil.apply_stencil(Stencil.stencil, arr, i, j, base, operator);
            }
        }
        // Edges
        // Top and bottom
        for (var j = 1; j < ncols - 1; j++) {
            //            console.log('top');
            new_arr[0][j] = Stencil.apply_stencil(Stencil.stencil_top, arr, 0, j, base, operator);
            //           console.log('bottom');
            new_arr[nrows - 1][j] = Stencil.apply_stencil(Stencil.stencil_bottom, arr, nrows - 1, j, base, operator);
        }
        // Left and right
        for (var i = 1; i < nrows - 1; i++) {
            new_arr[i][0] = Stencil.apply_stencil(Stencil.stencil_left, arr, i, 0, base, operator);
            new_arr[i][ncols - 1] = Stencil.apply_stencil(Stencil.stencil_right, arr, i, ncols - 1, base, operator);
        }
        // Corners
        new_arr[0][0] = Stencil.apply_stencil(Stencil.stencil_topleft, arr, 0, 0, base, operator);
        new_arr[0][ncols - 1] = Stencil.apply_stencil(Stencil.stencil_topright, arr, 0, ncols - 1, base, operator);
        new_arr[nrows - 1][0] = Stencil.apply_stencil(Stencil.stencil_bottomleft, arr, nrows - 1, 0, base, operator);
        new_arr[nrows - 1][ncols - 1] = Stencil.apply_stencil(Stencil.stencil_bottomright, arr, nrows - 1, ncols - 1, base, operator);
        return new_arr;
    };
    Stencil.reflectStencils = true;
    Stencil.ninePointStencil = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 0], [0, 1], [1, -1], [1, 0], [1, 1]];
    Stencil.stencil = Stencil.ninePointStencil;
    Stencil.initialized = false;
    return Stencil;
}());
exports.Stencil = Stencil;
